packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,176 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-      if (plugin.pluginDartClassPlatforms[platformKey] == null ||
-          plugin.pluginDartClassPlatforms[platformKey] == 'none') {
-        continue;
+      if (implementsPluginName != null) {
+        pluginImplCandidates.putIfAbsent(implementsPluginName, () => <Plugin>[]);
+        pluginImplCandidates[implementsPluginName]!.add(plugin);
       }
-
-      // If it hasn't been skipped, it's a candidate for auto-registration, so
-      // add it as a possible resolution.
-      possibleResolutions.putIfAbsent(implementsPackage, () => <Plugin>[]);
-      possibleResolutions[implementsPackage]!.add(plugin);
     }
 
-    final List<Plugin> pluginResolution = <Plugin>[];
+    final Map<String, Plugin> pluginResolution = <String, Plugin>{};
 
-    // Resolve all the possible resolutions to a single option for each plugin, or throw if that's not possible.
-    for (final MapEntry<String, List<Plugin>> entry in possibleResolutions.entries) {
-      final List<Plugin> candidates = entry.value;
-      // If there's only one candidate, use it.
-      if (candidates.length == 1) {
-        pluginResolution.add(candidates.first);
-        continue;
-      }
-      // Next, try direct dependencies of the resolving application.
-      final Iterable<Plugin> directDependencies = candidates.where((Plugin plugin) {
-        return plugin.isDirectDependency;
-      });
-      if (directDependencies.isNotEmpty) {
-        if (directDependencies.length > 1) {
-          globals.printError(
-              'Plugin ${entry.key}:$platformKey has conflicting direct dependency implementations:\n'
-                  '${directDependencies.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                  'To fix this issue, remove all but one of these dependencies from pubspec.yaml.\n'
-          );
-          hasResolutionError = true;
-        } else {
-          pluginResolution.add(directDependencies.first);
+    // Now resolve all the possible resolutions to a single option for each
+    // plugin, or throw if that's not possible.
+    for (final MapEntry<String, List<Plugin>> implCandidatesEntry in pluginImplCandidates.entries) {
+      try {
+        final Plugin? resolution = _resolveImplementationOfPlugin(
+          platformKey,
+          implCandidatesEntry,
+          defaultPackageName: defaultImplementations[implCandidatesEntry.key],
+        );
+        if (resolution != null) {
+          pluginResolution[implCandidatesEntry.key] = resolution;
         }
-        continue;
-      }
-      // Next, defer to the default implementation if there is one.
-      final String? defaultPackageName = defaultImplementations[entry.key];
-      if (defaultPackageName != null) {
-        final int defaultIndex = candidates
-            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
-        if (defaultIndex != -1) {
-          pluginResolution.add(candidates[defaultIndex]);
-          continue;
+      } on ToolExit catch (e) {
```

Comment:
Why are you catching `ToolExit`? It's supposed to be fatal. If you want to pass expected error messages around that are not final, there should be an explicit mechanism for that, not internal throw/catch.

Reply:
Sure. Can you point me to an example, how this is done elsewhere? The main reason for this is, that all errors e.g. for concurrent plugin implemntations had to be printed, before actually throwing (for all together). This means, it's simple to print them immediately, but one needs to decide if each plugin can be resolved correctly, before actually leaving the application. The error message itself keeps the same.

Reply:
Thanks, done in c1991f5393795b7e4dc9e1afbc68b3735260045e